TeamCode/src/main/java/org/firstinspires/ftc/teamcode/PixyCam.java

```diff
@@ -0,0 +1,170 @@
+ * Created by NoahR on 10/3/17.
+ * Note about Modern Robotics I2C wiring. (Left to right)
+ * 0 - 1 - 2 - 3
+ * 5v - Data - Clock - gnd
+ */
+
+public class PixyCam
+{
+    I2cDeviceSynch m_deviceClient;
+    private byte[] m_registerValues;
+    private int m_dataStartAddress = 0;
+
+    private final int START_ADDRESS = 0x54;
+    private final int READ_LENGTH = 26; // Start read at 0x58, end at 0x58 + 26
+    private final byte[] PIXY_DEVICE_SIGNATURE = {85, (byte) 170, 85, (byte) 170};
+    private final int SIGNATURE_SPACE_SIZE = 10; // Device signature will be in the first 10 bytes
+
+    PixyCam() {}
+    PixyCam( I2cDeviceSynch deviceClient ) { initialize( deviceClient); }
+
+    public void initialize( I2cDeviceSynch deviceClient )
+    {
+        m_deviceClient = deviceClient;
+        m_deviceClient.engage();
+    }
+
+    public byte[] GetRegisterValues() { return m_registerValues; }
+
+    /**
+     * Update the cashed register values
+     * @return whether the operation was successful
+     */
+    public boolean Update()
+    {
+
+        byte[] registerValues = m_deviceClient.read( START_ADDRESS, READ_LENGTH );
+        byte[] signatureSpace = new byte[SIGNATURE_SPACE_SIZE];
+
+        // Only use the first ten bytes for signature scanning
+        for( int iIndex = 0; iIndex < SIGNATURE_SPACE_SIZE; iIndex++)
+        {
+            signatureSpace[iIndex] = (byte) mod(registerValues[iIndex], 256);
+        }
+
+        // If all signature space values are zero then we are not retrieving valid data.
+        int signatureIndex = indexOfSubArray(signatureSpace, PIXY_DEVICE_SIGNATURE);
+        if( signatureIndex > -1 )
+        {
+            m_registerValues = registerValues;
+            m_dataStartAddress = signatureIndex;
+            return true;
+        }
+        else
+            return false;
+    }
+
+    /**
+     * Update the register values. If the values are not available, halt the thread until they are.
+     */
+    public void ForceUpdate()
+    {
+        while(!Update());
+    }
+
+    public int GetDataStartAddress()
+    {
+        return m_dataStartAddress;
+    }
+
+    /**
+     * @return the X coordinate of the center of the found object. Returns -1 if no data is
+     * available from the pixy cam
+     */
+    public int GetXCenter() { return ReadSignedInt(PixyDataOffsets.X_CENTER);}
+
+    /**
+     * @return the Y coordinate of the center of the found object. Returns -1 if no data is
+     * available from the pixy cam
+     */
+    public int GetYCenter() { return ReadSignedInt(PixyDataOffsets.Y_CENTER); }
+
+    /**
+     * @return the width of the found object. Returns -1 if no data is available from the pixy cam.
+     */
+    public int GetWidth() { return ReadSignedInt(PixyDataOffsets.WIDTH); }
+
+    /**
+     * @return the height of the found object. Returns -1 if no data is available from the pixy cam.
+     */
+    public int GetHeight() { return ReadSignedInt(PixyDataOffsets.HEIGHT); }
+
```

Comment:
There's a lot of heroic bit shifting and basic computer science exercise code below. It's great that you understand all this! That also means that (a) it's too general to go in this class, and (b) until you write your own tests, it's likely that a standard class does all this stuff already, and it's easier to use and maintain.

In this case, I would recommend using the [`ByteBuffer`](https://docs.oracle.com/javase/6/docs/api/java/nio/ByteBuffer.html) class, which is meant to wrap a byte array and correctly parse multi-byte fields to other primitive types in java. You may still have to do some unsigned arithmetic yourself, but this will do most of the heavy lifting for you.

    ByteBuffer dataBlock = ByteBuffer::wrap(m_deviceClient.read( START_ADDRESS, READ_LENGTH ));
    dataBlock.order(ByteOrder::LITTLE_ENDIAN); // See PixyCam spec
    int centerX = dataBlock.getShort(PixyDataOffsets::X_CENTER));
    ...